### PR TITLE
Fix IOperation APIs for object and collection initializers

### LINKED ISSUE #18781: IOperation API shape for object and collection initializer expressions
Currently, BoundCollectionIntializerExpression has OperationKind.None and implements IOperation. This means that we completely cut-off the operations within a collection initializer, including even the nested initializers.

We need to design the API shape for Collection initializers.

1. Do we need an `ICollectionInitializerExpression`? Theoretically, we don't need it as `IObjectCreationExpression.Initializers` can return the child collection element initializer nodes of a collection initializer.
2. Do we need an `ICollectionElementInitializerExpression`? We have 2 possible designs:
   1. Follow the VB bound tree pattern, which stores a `BoundCall` for collection element initializer, so we don't need a new API. However, C# doesn't store this bound call, but instead has a `BoundCollectionElementInitializerExpression` (with invoked method name and arguments) and `BoundDynamicCollectionElementInitializer` (with possible applicable methods and arguments) . So we need to synthesize such a bound call for C# and return it as an initializer. Note that this approach introduces an implicit compiler generated IOperation node in the operation tree.
   2. Add the following API and our Operation tree node for CollectionElementInitializerExpression, can have language specific implementations that delegate to either BoundCall (VB) or BoundICollectionElementInitializerExpression or BoundDynamicCollectionElementInitializer (C#):
```
namespace Microsoft.CodeAnalysis.Semantics
{
    /// <summary>
    /// Represents a collection element initializer expression.
    /// </summary>
    /// <remarks>
    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
    /// change it in the future.
    /// </remarks>
    public interface ICollectionElementInitializerExpression : IHasArgumentsExpression
    {
        /// <summary>
        /// Set of applicable methods for an implicit dynamic invocation OR the implicit Add method symbol for non-dynamic invocation.
        /// </summary>
        ImmutableArray<IMethodSymbol> ApplicableMethods { get; }

        /// <summary>
        /// Flag indicating if this is a dynamic initializer invocation.
        /// </summary>
        bool IsDynamic { get; }
    }
}
```

look for an example here  https://github.com/dotnet/roslyn/issues/18509

### PR DESCRIPTION
**Pull request by mavasani**
See https://github.com/dotnet/roslyn/issues/18781#issuecomment-311428609 for the final design.

Fixes #18781

### CODE REVIEW

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 1]
```diff
@@ -1329,6 +1329,10 @@
     <Field Name="Expanded" Type="bool"/>
     <Field Name="ArgsToParamsOpt" Type="ImmutableArray&lt;int&gt;" Null="allow"/>
     <Field Name="ResultKind" PropertyOverrides="true" Type="LookupResultKind"/>
+    
+    <!-- BinderOpt and UseSetterForDefaultArgumentGeneration are added as a temporary solution for IOperation implementation and should probably be removed in the future -->
```
Comment (mavasani) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: This is similar to the current approach used for BoundIndexerAccess

See https://github.com/mavasani/roslyn/blob/07421be2f2618cae7226438c06401fe0a40d7553/src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml#L1505. @genlu for any context.

**Comment by mavasani** [comment]
Also tagging @dotnet/analyzer-ioperation for review

**Comment by CyrusNajmabadi** [comment]
👍

**Comment by jinujoseph** [comment]
cc/ @dotnet/roslyn-compiler

**Comment by mavasani** [comment]
Ping.. need compiler signoff on the PR.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 2]
```diff
@@ -400,7 +410,100 @@ private IObjectCreationExpression CreateBoundObjectCreationExpressionOperation(B
...
+            bool isInvalid = boundObjectInitializerMember.HasErrors;
+            SyntaxNode syntax = boundObjectInitializerMember.Syntax;
+            ITypeSymbol type = boundObjectInitializerMember.Type;
+            Optional<object> constantValue = ConvertToOptional(boundObjectInitializerMember.ConstantValue);
+
+            switch (boundObjectInitializerMember.MemberSymbol.Kind)
+            {
+                case SymbolKind.Field:
+                    var field = (FieldSymbol)boundObjectInitializerMember.MemberSymbol;
+                    return new LazyFieldReferenceExpression(field, instance, field, isInvalid, syntax, type, constantValue);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Is the second `field` parameter necessary for `LazyFieldReferenceExpression`? (Are the two parameter values ever different?) Same question for `LazyEventReferenceExpression` and `LazyPropertyReferenceExpression`.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 3]
```diff
@@ -400,7 +410,100 @@ private IObjectCreationExpression CreateBoundObjectCreationExpressionOperation(B
...
+                    }
+                    else
+                    {
+                        // Indexed property reference.
+                        argumentsInEvaluationOrder = new Lazy<ImmutableArray<IArgument>>(() =>
+                        {
+                            return DeriveArguments(boundObjectInitializerMember.HasErrors,
+                                boundObjectInitializerMember.BinderOpt,
+                                property,
+                                property.SetMethod,
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: `CreateBoundIndexerAccessOperation()` uses `GetOwnOrInheritedSetMethod()`. Are there cases where that will give a different accessor here? If so, consider extracting a helper method that can be used by the factory methods.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 1 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: Thanks for the context.
Do we have an issue tracking removing this in the future?

---
In reply to: [126044205](https://github.com/dotnet/roslyn/pull/20689#discussion_r126044205) [](ancestors = 126044205)

Comment (mavasani) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: @genlu?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 4]
```diff
@@ -4008,8 +4008,10 @@ private BoundExpression MakeBadExpressionForObjectCreation(ObjectCreationExpress
                 expanded,
                 argsToParamsOpt,
                 resultKind,
-                boundMember.Type,
-                hasErrors);
+                binderOpt: this,
+                useSetterForDefaultArgumentGeneration: true,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: >useSetterForDefaultArgumentGeneration [](start = 16, length = 37)

Is there any case where `useSetterForDefaultArgumentGeneration` is `false`?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 5]
```diff
@@ -237,7 +247,7 @@ private IInvocationExpression CreateBoundCallOperation(BoundCall boundCall)
                         boundCall.ReceiverOpt != null &&
                         (boundCall.Method.IsVirtual || boundCall.Method.IsAbstract || boundCall.Method.IsOverride) &&
                         !boundCall.ReceiverOpt.SuppressVirtualCalls;
-            Lazy<ImmutableArray<IArgument>> argumentsInEvaluationOrder = new Lazy<ImmutableArray<IArgument>>(() => DeriveArguments(boundCall, boundCall.BinderOpt, boundCall.Method, boundCall.Method, boundCall.Arguments, boundCall.ArgumentNamesOpt, boundCall.ArgsToParamsOpt, boundCall.ArgumentRefKindsOpt, boundCall.Method.Parameters, boundCall.Expanded, boundCall.Syntax, boundCall.InvokedAsExtensionMethod));
+            Lazy<ImmutableArray<IArgument>> argumentsInEvaluationOrder = new Lazy<ImmutableArray<IArgument>>(() => DeriveArguments(boundCall.HasErrors, boundCall.BinderOpt, boundCall.Method, boundCall.Method, boundCall.Arguments, boundCall.ArgumentNamesOpt, boundCall.ArgsToParamsOpt, boundCall.ArgumentRefKindsOpt, boundCall.Method.Parameters, boundCall.Expanded, boundCall.Syntax, boundCall.InvokedAsExtensionMethod));
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Break line

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 6]
```diff
@@ -400,7 +410,100 @@ private IObjectCreationExpression CreateBoundObjectCreationExpressionOperation(B
...
+            bool isInvalid = boundCollectionInitializerExpression.HasErrors;
+            SyntaxNode syntax = boundCollectionInitializerExpression.Syntax;
+            ITypeSymbol type = boundCollectionInitializerExpression.Type;
+            Optional<object> constantValue = ConvertToOptional(boundCollectionInitializerExpression.ConstantValue);
+            return new LazyObjectOrCollectionInitializerExpression(initializers, isInvalid, syntax, type, constantValue);
+        }
+
+        private IMemberReferenceExpression CreateBoundObjectInitializerMemberOperation(BoundObjectInitializerMember boundObjectInitializerMember)
+        {
+            Lazy<IOperation> instance = new Lazy<IOperation>(() => new InstanceReferenceExpression(InstanceReferenceKind.Implicit, boundObjectInitializerMember.HasErrors, syntax: boundObjectInitializerMember.Syntax, type: boundObjectInitializerMember.MemberSymbol.ContainingType, constantValue: default(Optional<object>)));
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: break line
In general, we don't have a strict limit, but I'm confused why some code above (for example, call to `DeriveArguments`) is on multiple lines (good, readable), but some other lines are crazy long...

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 2 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Seems like it is redundant and there are more than one Operation types with such redundant parameters. I have added this to the list of comments to be addressed in the operation node generator: https://github.com/dotnet/roslyn/issues/20204#issue-235774159

Tagging @heejaechang as an FYI

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 4 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: This code path is while binding left of object member initializer, so it should always invoke the setter.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 3 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Thanks, will change to use that method here as well.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs [hunk 7]
```diff
@@ -73,7 +73,7 @@ internal static IArgument CreateArgumentOperation(ArgumentKind kind, IParameterS
             //      Note this check doesn't cover all scenarios. For example, when a parameter is a generic type but the type of the type argument 
             //      is undefined.
             if ((object)optionalParametersMethod == null 
-                || boundNode.HasAnyErrors
+                || hasErrors
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs]: It looks like callers are now passing `boundNode.HasErrors` rather than `boundNode.HasAnyErrors`. I'm not sure what the implications are but was that change intentional?

File: src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs [hunk 8]
```diff
@@ -508,10 +677,21 @@ System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
...
+      Initializers(1): IInvocationExpression ( ? MyList<System.String>.Add()) (OperationKind.InvocationExpression, Type: ?, IsInvalid) (Syntax: '""str""')
+          Instance Receiver: IInstanceReferenceExpression (InstanceReferenceKind.Implicit) (OperationKind.InstanceReferenceExpression, Type: MyList<System.String>) (Syntax: 'MyList<string>')
+          Arguments(1): IArgument (ArgumentKind.Explicit, Matching Parameter: null) (OperationKind.Argument, IsInvalid) (Syntax: '""str""')
+              ILiteralExpression (OperationKind.LiteralExpression, Type: System.String, Constant: ""str"") (Syntax: '""str""')
+";
+            var expectedDiagnostics = new DiagnosticDescription[] {
+                // CS1061: 'MyList<string>' does not contain a definition for 'Add' and no extension method 'Add' accepting a first argument of type 'MyList<string>' could be found (are you missing a using directive or an assembly reference?)
+                //         var coll = /*<bind>*/new MyList<string> { "str" }/*</bind>*/;
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, @"""str""").WithArguments("MyList<string>", "Add").WithLocation(7, 51)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs]: The location was dropped from the comment, here and in other cases below.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 4 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: I'm suggesting `useSetterForDefaultArgumentGeneration` should not exist if it is always `true`. I could not find code that could set it to `false`, but we should double-check.

File: src/Compilers/Core/Portable/Operations/ICollectionElementInitializerExpression.cs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    /// nested collection element initializer for Y, i.e. "{ { x, y, 3 } }" is represented by this operation.
+    /// </summary>
+    /// <remarks>
+    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
+    /// change it in the future.
+    /// </remarks>
+    public interface ICollectionElementInitializerExpression : IOperation
+    {
+        /// <summary>
+        /// Add method invoked on collection. Might be null for dynamic invocation.
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ICollectionElementInitializerExpression.cs]: It looks like

File: src/Compilers/Core/Portable/Operations/ICollectionElementInitializerExpression.cs [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        /// </summary>
+        IMethodSymbol AddMethod { get; }
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ICollectionElementInitializerExpression.cs]: It looks like `AddMethod` is always `null` for dynamic invocation. Consider changing comment from `Might be null ...`.

File: src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb [hunk 11]
```diff
@@ -498,6 +502,40 @@ Namespace Microsoft.CodeAnalysis.Semantics
...
+            Dim initializers As Lazy(Of ImmutableArray(Of IOperation)) = New Lazy(Of ImmutableArray(Of IOperation))(Function() boundCollectionInitializerExpression.Initializers.SelectAsArray(Function(n) CreateBoundCollectionElementInitializerOperation(n)))
+            Dim isInvalid As Boolean = boundCollectionInitializerExpression.HasErrors
+            Dim syntax As SyntaxNode = boundCollectionInitializerExpression.Syntax
+            Dim type As ITypeSymbol = boundCollectionInitializerExpression.Type
+            Dim constantValue As [Optional](Of Object) = ConvertToOptional(boundCollectionInitializerExpression.ConstantValueOpt)
+            Return New LazyObjectOrCollectionInitializerExpression(initializers, isInvalid, syntax, type, constantValue)
+        End Function
+
+        Private Function CreateBoundCollectionElementInitializerOperation(boundExpression As BoundExpression) As IOperation
+            If (boundExpression.Kind <> BoundKind.Call) Then
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb]: Parentheses are not needed.

File: src/Test/Utilities/Portable/Compilation/TestOperationWalker.cs [hunk 12]
```diff
@@ -421,6 +421,24 @@ public override void VisitAnonymousObjectCreationExpression(IAnonymousObjectCrea
...
+        }
+
+        public override void VisitMemberInitializerExpression(IMemberInitializerExpression operation)
+        {
+            base.VisitMemberInitializerExpression(operation);
+        }
+
+        public override void VisitCollectionElementInitializerExpression(ICollectionElementInitializerExpression operation)
+        {
+            var addMethod = operation.AddMethod;
```
Comment (jcouv) [inline src/Test/Utilities/Portable/Compilation/TestOperationWalker.cs]: Two unused locals

**Comment by jcouv** [comment]
>             var ctor = operation.Constructor;

Unused local

---
Refers to: src/Test/Utilities/Portable/Compilation/TestOperationWalker.cs:414 in 827341b. [](commit_id = 827341bfa471863b90a5d408c8503d45629d9b37, deletion_comment = False)

File: src/Test/Utilities/Portable/Compilation/TestOperationWalker.cs (hunk 12 above)
Comment (jcouv) [inline src/Test/Utilities/Portable/Compilation/TestOperationWalker.cs]: The behavior of this override and a few above doesn't add anything to the base behavior. They can be removed.

---
In reply to: [126534989](https://github.com/dotnet/roslyn/pull/20689#discussion_r126534989) [](ancestors = 126534989)

**Comment by jcouv** [comment]
>             var text = operation.Text;

Seems a recurring pattern (unused local, override only calls base). Am I missing something?

---
Refers to: src/Test/Utilities/Portable/Compilation/TestOperationWalker.cs:397 in 827341b. [](commit_id = 827341bfa471863b90a5d408c8503d45629d9b37, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs [hunk 13]
```diff
@@ -57,50 +57,64 @@ public void M1()
...
   IVariableDeclarationStatement (1 declarations) (OperationKind.VariableDeclarationStatement, IsInvalid) (Syntax: 'var e1 = ne ... rty2 = 1 };')
     IVariableDeclaration (1 variables) (OperationKind.VariableDeclaration, IsInvalid) (Syntax: 'var e1 = ne ... rty2 = 1 };')
       Variables: Local_1: F e1
       Initializer: IObjectCreationExpression (Constructor: F..ctor()) (OperationKind.ObjectCreationExpression, Type: F, IsInvalid) (Syntax: 'new F() { P ... erty2 = 1 }')
-          Initializers(1): ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: B, IsInvalid) (Syntax: 'Property2 = 1')
-              Left: IOperation:  (OperationKind.None) (Syntax: 'Property2')
-              Right: IConversionExpression (ConversionKind.Invalid, Implicit) (OperationKind.ConversionExpression, Type: B, IsInvalid) (Syntax: '1')
-                  ILiteralExpression (Text: 1) (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1) (Syntax: '1')
+          Initializer: IObjectOrCollectionInitializerExpression (OperationKind.ObjectOrCollectionInitializerExpression, Type: F, IsInvalid) (Syntax: '{ Property2 = 1 }')
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: Nit: I raised this question earlier, but I don't remember if it was answered. Do we know why some indents are 2 spaces, but others are 4?

**Comment by mavasani** [comment]
I believe @genlu wrote this walker as a checker that runs on any code and verifies no crashes on accessing properties off operation nodes.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs (hunk 13 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: Some formatting will be fixed by Jinu's PR, but probably not this issue.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 4 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Sure, let me make that change.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs (hunk 7 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs]: I'll change it to HasAnyErrors at all call sites, it was not an intentional change.

File: src/Compilers/Core/Portable/Operations/ICollectionElementInitializerExpression.cs (hunk 10 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/Operations/ICollectionElementInitializerExpression.cs]: Sure.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs (hunk 7 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs]: Requiring callers to pass `HasAnyErrors` seems error prone. Is it possible to revert to passing in the `BoundNode` instead?

Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs]: Yes, that seems better.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 6 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: I will add break lines for the couple of cases here. The formatting is not consistent across this file, so I have added a comment to https://github.com/dotnet/roslyn/issues/20204#issue-235774159. Tagging @heejaechang as an FYI, we should address this in the operation node generator.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 1 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: Filed https://github.com/dotnet/roslyn/issues/20787

File: src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs (hunk 8 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs]: Operation unit test generator doesn't seem to include the diagnostic location in the generated comment for expected diagnostics. Filed https://github.com/dotnet/roslyn/issues/20788 to track updating the generator.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs (hunk 13 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs]: Filed https://github.com/dotnet/roslyn/issues/20789 to track the indenting issues.

**Comment by mavasani** [comment]
@jcouv @cston - addressed your feedback.

**Comment by cston** [comment]
LGTM

**Comment by mavasani** [comment]
windows_debug_unit32_prtest failed due to flaky test failure tracked by https://github.com/dotnet/roslyn/issues/14592

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit32_prtest please

**Comment by mavasani** [comment]
windows_debug_unit64_prtest failed due to flaky test https://github.com/dotnet/roslyn/issues/20370

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please
